server/src/test/java/org/elasticsearch/index/mapper/TextFieldMapperTests.java

```diff
@@ -956,4 +963,101 @@ public void testIndexPrefixMapping() throws IOException {
             assertThat(e.getMessage(), containsString("Cannot set index_prefixes on unindexed field [field]"));
         }
     }
+
+    public void testFastPhrasePrefixes() throws IOException {
+        QueryShardContext queryShardContext = indexService.newQueryShardContext(
+            randomInt(20), null, () -> {
+                throw new UnsupportedOperationException();
+            }, null);
+
+        String mapping = Strings.toString(XContentFactory.jsonBuilder().startObject().startObject("type")
+            .startObject("properties")
+                .startObject("field")
+                    .field("type", "text")
+                    .field("analyzer", "my_stop_analyzer")
+                    .startObject("index_prefixes")
+                        .field("min_chars", 2)
+                        .field("max_chars", 10)
+                    .endObject()
+                .endObject()
+                .startObject("synfield")
+                    .field("type", "text")
+                    .field("analyzer", "standard")  // will be replaced with MockSynonymAnalyzer
+                    .field("index_phrases", true)
+                    .startObject("index_prefixes")
+                        .field("min_chars", 2)
+                        .field("max_chars", 10)
+                    .endObject()
+                .endObject()
+            .endObject()
+            .endObject().endObject());
+
+        queryShardContext.getMapperService().merge("type", new CompressedXContent(mapping), MergeReason.MAPPING_UPDATE);
+
+        Query q = new MatchPhrasePrefixQueryBuilder("field", "two words").toQuery(queryShardContext);
+        Query expected = new SpanNearQuery.Builder("field", true)
+            .addClause(new SpanTermQuery(new Term("field", "two")))
+            .addClause(new FieldMaskingSpanQuery(
+                new SpanTermQuery(new Term("field._index_prefix", "words")), "field")
+            )
+            .build();
+        assertThat(q, equalTo(expected));
+
+        Query q2 = new MatchPhrasePrefixQueryBuilder("field", "three words here").toQuery(queryShardContext);
+        expected = new SpanNearQuery.Builder("field", true)
+            .addClause(new SpanTermQuery(new Term("field", "three")))
+            .addClause(new SpanTermQuery(new Term("field", "words")))
+            .addClause(new FieldMaskingSpanQuery(
+                new SpanTermQuery(new Term("field._index_prefix", "here")), "field")
+            )
+            .build();
+        assertThat(q2, equalTo(expected));
+
+        Query q3 = new MatchPhrasePrefixQueryBuilder("field", "two words").slop(1).toQuery(queryShardContext);
+        expected = new SpanNearQuery.Builder("field", true)
+            .setSlop(1)
```

Comment:
Does this change the semantics of slop of a match phrase prefix?  Slop on a phrase query includes swapped positions, while slop on a span near only includes gaps.

Reply:
Yes good catch, I modified the code so that we don't apply the optimization when the `slop` is greater than 0.